evm/src/cpu/kernel/asm/memory/core.asm

```diff
@@ -393,3 +393,21 @@
     %mstore_kernel_general_2
     // stack: (empty)
 %endmacro
+
+%macro mload_main
+    // stack: offset
+    DUP1
+    // stack: offset, offset
+    %update_msize
+    // stack: offset
+    %mload_current(@SEGMENT_MAIN_MEMORY)
+%endmacro
+
+%macro mstore_main
+    // stack: offset, value
+    DUP1
+    // stack: offset, offset, value
+    %update_msize
+    // stack: offset, value
+    %mstore_current(@SEGMENT_MAIN_MEMORY)
+%endmacro
```

Comment:
One drawback of this approach is that we need to make sure we always use these macros when touching main memory.

Reply:
I think it seems fine, we can use these to implement syscalls for MLOAD/MSTORE, and then for the few other opcodes that affect memory size (like Keccak) we can call `update_msize` directly